File: bryannet-admin/src/api/devices.js
import api from "./axios";

/**
 * Get all devices.
 */
export const getDevices = async () => {

    const response = await api.get(
        "/devices",
    );

    return response.data.items;

};

/**
 * Get devices belonging to a customer.
 */
export const getCustomerDevices = async (
    customerId,
) => {

    const response = await api.get(
        `/devices/customer/${customerId}`,
    );

    return response.data;

};

/**
 * Register a new device.
 */
export const registerDevice = async (
    deviceData,
) => {

    const response = await api.post(
        "/devices/register",
        deviceData,
    );

    return response.data;

};

/**
 * Activate a device.
 */
export const activateDevice = async (
    deviceId,
) => {

    const response = await api.patch(
        `/devices/${deviceId}/activate`,
    );

    return response.data;

};

/**
 * Deactivate a device.
 */
export const deactivateDevice = async (
    deviceId,
) => {

    const response = await api.patch(
        `/devices/${deviceId}/deactivate`,
    );

    return response.data;

};

/**
 * Approve a device.
 */
export const approveDevice = async (
    deviceId,
) => {

    const response = await api.patch(
        `/devices/${deviceId}/approve`,
    );

    return response.data;

};

/**
 * Block a device.
 */
export const blockDevice = async (
    deviceId,
) => {

    const response = await api.patch(
        `/devices/${deviceId}/block`,
    );

    return response.data;

};

/**
 * Unblock a device.
 */
export const unblockDevice = async (
    deviceId,
) => {

    const response = await api.patch(
        `/devices/${deviceId}/unblock`,
    );

    return response.data;

};

/**
 * Rename a device.
 */
export const renameDevice = async (
    deviceId,
    deviceName,
) => {

    const response = await api.patch(
        `/devices/${deviceId}/rename`,
        {
            device_name: deviceName,
        },
    );

    return response.data;

};

/**
 * Replace a device.
 */
export const replaceDevice = async (
    replacementData,
) => {

    const response = await api.post(
        "/devices/replace",
        replacementData,
    );

    return response.data;

};

File: app/portal/devices.py
from fastapi import (
    APIRouter,
    Depends,
)

from sqlalchemy.orm import (
    Session,
)

from app.database.dependencies import (
    get_db,
)

from app.schemas.portal_device import (
    PortalDeviceActionRequest,
    PortalDeviceListResponse,
)

from app.services.portal.device_service import (
    PortalDeviceService,
)

router = APIRouter(
    prefix="/devices",
    tags=["Portal Devices"],
)


# ==========================================================
# Query Methods
# ==========================================================

@router.get(
    "/{telegram_user_id}",
    response_model=PortalDeviceListResponse,
)
def get_customer_devices(
    telegram_user_id: int,
    db: Session = Depends(
        get_db,
    ),
):

    return (
        PortalDeviceService.get_customer_devices(
            db,
            telegram_user_id,
        )
    )


# ==========================================================
# Business Commands
# ==========================================================

@router.patch(
    "/{device_id}/rename",
    response_model=PortalDeviceListResponse,
)
def rename_device(
    device_id: int,
    request: PortalDeviceActionRequest,
    db: Session = Depends(
        get_db,
    ),
):

    return (
        PortalDeviceService.rename_device(
            db,
            device_id,
            request,
        )
    )


@router.patch(
    "/{device_id}/activate",
    response_model=PortalDeviceListResponse,
)
def activate_device(
    device_id: int,
    request: PortalDeviceActionRequest,
    db: Session = Depends(
        get_db,
    ),
):

    return (
        PortalDeviceService.activate_device(
            db,
            device_id,
            request,
        )
    )


@router.patch(
    "/{device_id}/deactivate",
    response_model=PortalDeviceListResponse,
)
def deactivate_device(
    device_id: int,
    request: PortalDeviceActionRequest,
    db: Session = Depends(
        get_db,
    ),
):

    return (
        PortalDeviceService.deactivate_device(
            db,
            device_id,
            request,
        )
    )


@router.patch(
    "/{device_id}/block",
    response_model=PortalDeviceListResponse,
)
def block_device(
    device_id: int,
    request: PortalDeviceActionRequest,
    db: Session = Depends(
        get_db,
    ),
):

    return (
        PortalDeviceService.block_device(
            db,
            device_id,
            request,
        )
    )


@router.patch(
    "/{device_id}/unblock",
    response_model=PortalDeviceListResponse,
)
def unblock_device(
    device_id: int,
    request: PortalDeviceActionRequest,
    db: Session = Depends(
        get_db,
    ),
):

    return (
        PortalDeviceService.unblock_device(
            db,
            device_id,
            request,
        )
    )